randomwallpaper@iflow.space/sourceAdapter.js

```diff
@@ -356,13 +356,61 @@ var GenericJsonAdapter = class extends BaseAdapter {
 			try {
 				const response_body = JSON.parse(ByteArray.toString(response_body_bytes));
 
-				let JSONPath = this._settings.get("generic-json-response-path", "string");
-				let imageDownloadUrl = this._jsonPathParser.access(response_body, JSONPath);
-				imageDownloadUrl = this._settings.get("generic-json-url-prefix", "string") + imageDownloadUrl;
+				let identifier = this._settings.get("generic-json-id", "string");
+				let imageJSONPath = this._settings.get("generic-json-image-path", "string");
+				let postJSONPath = this._settings.get("generic-json-post-path", "string");
+				let domainUrl = this._settings.get("generic-json-domain", "string");
+				let authorNameJSONPath = this._settings.get("generic-json-author-name-path", "string");
+				let authorUrlJSONPath = this._settings.get("generic-json-author-url-path", "string");
+
+				if (identifier === null || identifier === "") {
+					identifier = 'Generic JSON Source';
+				}
+
+				let rObject = this._jsonPathParser.access(response_body, imageJSONPath);
+				let imageDownloadUrl = this._settings.get("generic-json-image-prefix", "string") + rObject.Object;
+
+				// '@random' would yield different results so lets make sure the values stay
+				// the same as long as the path is identical
+				let samePath = imageJSONPath.substring(0, this.findFirstDifference(imageJSONPath, postJSONPath));
+
+				// count occurrences of '@random' to slice the array later
+				// https://stackoverflow.com/a/4009768
+				let occurrences = (samePath.match(/@random/g) || []).length;
+				let slicedRandomElements = rObject.RandomElements.slice(0, occurrences);
+
+				let postUrl = this._jsonPathParser.access(response_body, postJSONPath, slicedRandomElements, false).Object;
+				postUrl = this._settings.get("generic-json-post-prefix", "string") + postUrl;
+				if (typeof postUrl !== 'string' || !postUrl instanceof String) {
+					postUrl = null;
+				}
+
+				let authorName = this._jsonPathParser.access(response_body, authorNameJSONPath, slicedRandomElements, false).Object;
+				if (typeof authorName !== 'string' || !authorName instanceof String) {
+					authorName = null;
+				}
+
+				let authorUrl = this._jsonPathParser.access(response_body, authorUrlJSONPath, slicedRandomElements, false).Object;
+				authorUrl = this._settings.get("generic-json-author-url-prefix", "string") + authorUrl;
+				if (typeof authorUrl !== 'string' || !authorUrl instanceof String) {
+					authorUrl = null;
+				}
```

Comment:
For me, the generic JSON source didn't work and threw an error `BaseAdapter :: {"error":"Unexpected response. (TypeError: rObject.RandomElements is null)"}`
I didn't debug the issue. Should that already work, or are there fixes for this in later PRs?

Reply:
It should work! I'll have a look at it, maybe something got lost while transfering the commits between the branches.

Reply:
Hm, it works just fine here.

In the past I had the problem that the gnome extension app was overwriting my local folder on gnome shell restart - maybe that's interfering on your side?

Reply:
Hm, it worked after restting my stored settings with dconf...  That's somewhat unfortunate since I can't restore my previous state to test what caused the issue. We could change the schema id in the end so that other users do not run into the same issue with the next version.

Fyi, created an issue for tracking here: https://github.com/ifl0w/RandomWallpaperGnome3/issues/156